### PR TITLE
[SPARK-39914][SQL] Add DS V2 Filter to V1 Filter conversion

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add util methods to convert DS V2 Filter to V1 Filter.

### Why are the changes needed?
Provide convenient methods to convert V2 to V1 Filters. These methods can be used by [`SupportsRuntimeFiltering`](https://github.com/apache/spark/pull/36918/files#diff-0d3268f351817ca948e75e7b6641e5cc67c4d773c3234920a7aa62faf11f6c8e) and later be used by `SupportsDelete`

### Does this PR introduce _any_ user-facing change?
No. These are intended for internal use only

### How was this patch tested?
new tests

### CODE REVIEW

**Comment by huaxingao** [comment]
The GA failure doesn't seem relevant.

**Comment by huaxingao** [comment]
@cloud-fan Could you please take a look when you have time? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala [hunk 1]
```diff
@@ -43,6 +43,73 @@ private[sql] object PredicateUtils {
           Some(In(attribute, Array.empty[Any]))
         }
 
+      case "=" | "<=>" | ">" | "<" | ">=" | "<=" if predicate.children().length == 2 &&
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala:46]: Both `case "=" | "<=>" | ">" | "<" | ">=" | "<="` and `"STARTS_WITH" | "ENDS_WITH" | "CONTAINS"` use condition

```
if predicate.children().length == 2 &&
  predicate.children()(0).isInstanceOf[NamedReference] &&
  predicate.children()(1).isInstanceOf[LiteralValue[_]]
```
Is it necessary to extract a function?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala [hunk 2]
```diff
@@ -43,6 +43,73 @@ private[sql] object PredicateUtils {
...
+
+      case "ALWAYS_TRUE" | "ALWAYS_FALSE" if predicate.children().isEmpty =>
+        predicate.name() match {
+          case "ALWAYS_TRUE" => Some(AlwaysTrue())
+          case "ALWAYS_FALSE" => Some(AlwaysFalse())
+        }
+
+      case "AND" =>
+        val and = predicate.asInstanceOf[V2And]
+        Some(And(toV1(and.left()).get, toV1(and.right()).get))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala:103]: I think we should return None if any of the children can't be converted to v1.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala [hunk 3]
```diff
@@ -43,6 +54,83 @@ private[sql] object PredicateUtils {
           Some(In(attribute, Array.empty[Any]))
         }
 
+      case "=" | "<=>" | ">" | "<" | ">=" | "<=" if isValidBinaryPredicate =>
+        val attribute = predicate.children()(0).toString
+        val value = predicate.children()(1).asInstanceOf[LiteralValue[_]]
+        predicate.name() match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala:60]: nit:
```
val v1Value = CatalystTypeConverters.convertToScala(value.value, value.dataType)
predicate.name() match {
  case "=" => Some(EqualTo(attribute, v1Value))
  ...
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala [hunk 4]
```diff
@@ -43,6 +54,83 @@ private[sql] object PredicateUtils {
...
+          Some(Or(left.get, right.get))
+        } else if (left.nonEmpty) {
+          left
+        } else {
+          right
+        }
+
+      case "NOT" =>
+        val not = predicate.asInstanceOf[V2Not]
+        Some(Not(toV1(not.child()).get))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/PredicateUtils.scala:132]: `toV1(not.child())` may not always return `Some`

**Review by cloud-fan — APPROVED** [review]

**Comment by huaxingao** [comment]
Thanks for reviewing! Merged to master.

**Pull request closed without merging**